File: src/maps2labels.m
function [ Label,Label2, Images,list] = maps2labels( I,L )
%converts indexes of layer contours to label images
    Label = [];
    Label2 = [];
    Images = [];
    list = [];
    for scan = 1:size(I,3)
        image = I(:,:,scan);
%         image = imresize(image,0.5);
        s = size(image);
        l = L(:,:,scan);
%         l = round(l/2);
%         l = imresize(l,[8,s(2)]);
        
        if sum(~isnan(sum(l)))>50
            list = [list;scan];
            for layer = 1:size(l,1)
                l_temp = l(layer,:);
                t = 1:numel(l_temp);
                l(layer,:) = interp1(t(~isnan(l_temp)),l_temp(~isnan(l_temp)),t,'linear','extrap');
            end
            l = round(l);
            l(l<3)=3;
            l(l>s(1)) = s(1);
            layers = ones(size(l,1)+1,s(2),2);
            layers(2:end,:,1) = l;
            layers(1:end-1,:,2) = l-1;
            layers(end,:,2) = s(1);
            label = zeros(s(1),s(2));
            label2 = zeros(s(1),s(2));
            for col = 1: s(2)
                for lay = 1: size(layers,1)
                    label(layers(lay,col,1):layers(lay,col,2),col)=lay;
                    label2(layers(lay,col,2),col)=lay;
%                     label(layers(lay,col,1):layers(lay,col,2),col)=lay;
%                     label2(layers(lay,col,1),col)=lay-1;
                end
            end
            Label = cat(3,Label,label);
            Label2 = cat(3,Label2,label2);
            Images = cat(3,Images, image);
        end
    end
end

File: src/maps2labels.py
import os
dr = '/home/mict/Chap_1/data'
os.chdir(dr)
os.getcwd() 
import glob
ls = sorted(glob.glob(dr+'/*.mat'))
import scipy.io as sio
import numpy as np
nan = np.nan
isnan = np.isnan

def interp_nan(layers):
    s=layers.shape
    L = np.zeros([s[0],s[1],s[2]])
#    L = zeros(size(layers));
    for temp in range(s[2]):
        #for temp = 1:size(layers,3)
        l = layers[:,:,temp]
        #l = layers(:,:,temp);
        for layer in range(s[0]):
        #for layer = 1:size(l,1)
            A = l[layer,:]
            ok = ~np.isnan(A)
            xp = ok.ravel().nonzero()[0]
            fp = A[~np.isnan(A)]
            x  = np.isnan(A).ravel().nonzero()[0]
            A[np.isnan(A)] = np.interp(x, xp, fp)
            l[layer,:] = A
        L[:,:,temp] = l
    return L

def maps2labels(I,L):
    list = []
    Label = np.zeros((I.shape[0],I.shape[1],I.shape[2]), dtype='uint8')
    Label2 = np.zeros((I.shape[0],I.shape[1],I.shape[2]), dtype='uint8')
    for scan in range(I.shape[2]):
        image = I[:,:,scan]
        s = image.shape
        l = L[:,:,scan]
        if np.sum(~np.isnan(l.sum(0)))>50:
            list.append(scan)
            l = np.round(l)
            l[l<3]=3
            l[l>s[0]] = s[0]
            layers = np.zeros((l.shape[0]+1,s[1],2), dtype='int')
            layers[1::,:,0] = l
            layers[:-1,:,1] = l
            layers[-1,:,1] = s[0]
            label = np.zeros((s[0],s[1]))
            label2 = np.zeros((s[0],s[1]))
            for col in range(s[1]):
                for lay in range(layers.shape[0]):
                    label[layers[lay,col,0]:layers[lay,col,1],col]=lay+1
                    label2[layers[lay,col,0],col]=lay
            Label[:,:,scan] = label
            Label2[:,:,scan] = label2
    return (Label,Label2,I,list)    

for x in range(len(ls)):
    #x=0
    data = sio.loadmat(ls[x])
    I = data['images']    #data.keys()
    L1 = data['manualLayers1']
    L2 = data['manualLayers2']
    nan_check = np.sum(np.sum(~isnan(L1),axis=0),axis=0)
    I = I[:,:,nan_check>0]
    Label1 = L1[:,:,nan_check>0]
    Label2 = L2[:,:,nan_check>0]
    L1 = interp_nan(Label1)
    L2 = interp_nan(Label2)
    L_M1,C_M1,Im,temp = maps2labels(I,L1-1) # as python is 0 index
    L_M2,C_M2,temp,temp = maps2labels(I,L2-1) # as python is 0 index
    L,C,temp,temp = maps2labels(I,0.5*(L1+L2)-1) # as python is 0 index
    if x==0:
        Labels = np.concatenate((L_M1,L_M2),axis=2)
        Contours = np.concatenate((C_M1,C_M2),axis=2)
        Images = np.concatenate((Im,Im),axis=2)
    else:
        Labels = np.concatenate((Labels,L_M1,L_M2),axis=2)
        Contours = np.concatenate((Contours,C_M1,C_M2),axis=2)
        Images = np.concatenate((Images,Im,Im),axis=2)
Labels = Labels[:,118:653,:]
Contours = Contours[:,118:653,:]
Images = Images[:,118:653,:]

import scipy.io as sio
sio.savemat('Data.mat', {'Images': Images, 'Label': Labels, 'Contour': Contours})
import matplotlib.pyplot as plt
imgplot = plt.imshow(Contours[:,:,0])    
